src/polygon.cpp

```diff
@@ -411,20 +409,9 @@ class EarClip {
       return true;
     }
 
-    // A major key to robustness is to only clip convex ears, but this is
-    // difficult to determine when an edge is folded back on itself. This
-    // function walks down the kinks in a degenerate portion of a polygon until
-    // it finds a clear geometric result. In the vast majority of cases the loop
-    // will only need one or two iterations.
+    // Returns true for convex or colinear ears.
     bool IsConvex(double epsilon) const {
-      const int convexity = CCW(left->pos, pos, right->pos, epsilon);
-      if (convexity != 0) {
-        return convexity > 0;
-      }
-      if (la::dot(left->pos - pos, right->pos - pos) <= 0) {
-        return true;
-      }
-      return left->InsideEdge(left->right, epsilon, true);
+      return CCW(left->pos, pos, right->pos, epsilon) >= 0;
```

Comment:
This is interesting - I had been operating under the assumption that I needed to effectively determine a correct perturbation of the epsilon-valid polygon to make it truly valid (non-self-intersecting). This was in fact necessary for my previous triangulator, but since ear clipping proceeds in a naturally manifold way, I tried just letting it clip all degenerate ears. It seems to work great, and vastly simplifies the code. I think the keyholing step may be the only part that's really sensitive to this kind of thing.